### PR TITLE
[TF][Relay] BatchNorm support with run-time mean and variance calculation

### PR DESCRIPTION
**Pull request by lfengad**
We observe a great amount of tensorflow models used in our production environment invoke the FusedBatchNorm operator, and a lot of them use this operator in "is_training" mode for model inference. In "is_training" mode, the mean and variance are calculated dynamically using the run-time data without pre-defined. However, the current BatchNorm in TVM requires the mean and variance are given as non-empty tensors. 

We add the support for BatchNorm in "is_training" mode, to make it able to dynamically calculate the mean and variance if not given. We first check the mean node and variance node for fused_batch_norm in tensorflow frontend to annotate them if they are empty. Then according to the annotation, we add necessary nodes for the mean and variance calculation in BatchNormToInferUnpack function, which is used to arrange the BatchNorm inference.

In our current implementation, the annotations of the empty mean and variance are added into the name_hint of the corresponding variable nodes. This solution is simple and no need to modify the attributes of the relay operator batch_norm. Alternatively, we can add a bool attribute "is_training" to the relay operator batch_norm. If the mean and variance are empty, "is_training" is set to true. Then according to the attributes of the relay operator, we decide whether to add the nodes for calculating the mean and variance or not in function BatchNormToInferUnpack. This solution needs to modify the relay operator batch_norm.

Any suggestions are welcome! @tqchen @FrozenGene

### CODE REVIEW

**Comment by FrozenGene** [comment]
I think we needn't add `_empty_for_training_mode_inference`. If we find `mean` / `variance` is `VarNode`, we should call `Mean` and `Variance`.

I don't think we should add `is_training` flag to relay `BatchNorm`. This should be done by users to make sure TF's model `BatchNorm`'s `is_training` flag be false. However, we still have user cases like you mention, so we could support as current implementation and don't add attribute to `BatchNorm`.

**Comment by lfengad** [comment]
> I think we needn't add `_empty_for_training_mode_inference`. If we find `mean` / `variance` is `VarNode`, we should call `Mean` and `Variance`.
> 
> I don't think we should add `is_training` flag to relay `BatchNorm`. This should be done by users to make sure TF's model `BatchNorm`'s `is_training` flag be false. However, we still have user cases like you mention, so we could support as current implementation and don't add attribute to `BatchNorm`.

Thank you so much for the quick reply!  
Yeah, our current implementation is just to check whether `mean` / `variance` is empty `VarNode` (with zero dimension),  and then call `Mean` and `Variance` in BatchNormToInferUnpack.  Also as I understand, if `mean` / `variance` is `VarNode` but with non-zero dimension, it still has the possibility to hold the given pre-defined constant values and thus cannot be replaced with `Mean` \ `Variance`.   
Thank you for the discussion!

**Comment by FrozenGene** [comment]
> Yeah, our current implementation is just to check whether `mean` / `variance` is empty `VarNode` (with zero dimension), and then call `Mean` and `Variance` in BatchNormToInferUnpack.

I think our pr could remove `name_hint` too.

>  if mean / variance is VarNode but with non-zero dimension, it still has the possibility to hold the given pre-defined constant values and thus cannot be replaced with Mean \ Variance.

Could you give us an example of this condition? I could only imagine models have empty or full pre-defined values. So we should only to calculate it by calling `Mean` / `Variance` feed by `data` or our current implementation of `BatchNormToInferUnpack `.

**Comment by lfengad** [comment]
> > Yeah, our current implementation is just to check whether `mean` / `variance` is empty `VarNode` (with zero dimension), and then call `Mean` and `Variance` in BatchNormToInferUnpack.
> 
> I think our pr could remove `name_hint` too.
Yeah, I agree that the better way should be removing `name_hint` and just checking whether the `mean` and `variance` are empty inside `BatchNormToInferUnpack`, with no need to modify the tensorflow frontend. Previously I have tried this way but got come compilation errors related with layout checking. If we plan to do in this way, we need to modify the layout checking of `batch_norm` operator too. 
> 
> > if mean / variance is VarNode but with non-zero dimension, it still has the possibility to hold the given pre-defined constant values and thus cannot be replaced with Mean \ Variance.
> 
> Could you give us an example of this condition? I could only imagine models have empty or full pre-defined values. So we should only to calculate it by calling `Mean` / `Variance` feed by `data` or our current implementation of `BatchNormToInferUnpack `.
What I mean is that for both cases the `mean` and `variance` are `VarNode`. In one case the `VarNode` is empty without pre-defined values, while in the other case the `VarNode` is not empty with pre-defined values. 
Thank you for the discussion!

**Comment by lfengad** [comment]
> > Yeah, our current implementation is just to check whether `mean` / `variance` is empty `VarNode` (with zero dimension), and then call `Mean` and `Variance` in BatchNormToInferUnpack.
> 
> I think our pr could remove `name_hint` too.
> 
> > if mean / variance is VarNode but with non-zero dimension, it still has the possibility to hold the given pre-defined constant values and thus cannot be replaced with Mean \ Variance.
> 
> Could you give us an example of this condition? I could only imagine models have empty or full pre-defined values. So we should only to calculate it by calling `Mean` / `Variance` feed by `data` or our current implementation of `BatchNormToInferUnpack `.

Thanks for your discussion! According to our discussion, I have rewritten the code as in the newest commit. This time, the function `BatchNormToInferUnpack` is not modified. We only modify the tensorflow frontend for `_fused_batch_norm`. If `mean` and `variance` are empty, we directly add `Mean` and `Variance` relay operators to the frontend graph before the `batch_norm` operator, without modifying the `batch_norm` operator at all.
Thank you for the suggestions!

File: tests/python/frontend/tensorflow/test_bn_trainingmod.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    layout = None
+    target = 'llvm'
+    ctx=tvm.cpu(0)
+    mod, params = relay.frontend.from_tensorflow(constant_graph, layout=layout, outputs=['sum'])
+    with relay.build_config(opt_level=3):
+        graph, lib, params = relay.build(mod,
+                                     target=target,
+                                     target_host = target,
+                                     params=params)
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_bn_trainingmod.py]: Align. Make sure `target=target` keep the same align `relay.build(mod,`

File: tests/python/frontend/tensorflow/test_bn_trainingmod.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    data = np.random.rand(1, 12, 12, 32)
+    with tf.Session(graph=out.graph) as sess:
+        sess.run([tf.global_variables_initializer()])
+        tf_out = sess.run(out, feed_dict={input_tensor:data})
+        constant_graph = graph_util.convert_variables_to_constants(sess, sess.graph_def, ['sum'])
+    
+
+    layout = None
+    target = 'llvm'
+    ctx=tvm.cpu(0)
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_bn_trainingmod.py]: space between `=`

File: tests/python/frontend/tensorflow/test_bn_trainingmod.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+This article is a test script to test fused_batch_norm operators in TensorFlow frontend when mean and variance are not given.
+"""
+import tvm
+import numpy as np
+import tensorflow as tf
+from tvm import relay
+from tensorflow.python.framework import graph_util
+
+def test_fusedbatchnorm():
+    g=tf.Graph()
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_bn_trainingmod.py]: space between `=`

File: tests/python/frontend/tensorflow/test_bn_trainingmod.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        graph, lib, params = relay.build(mod,
+                                     target=target,
+                                     target_host = target,
+                                     params=params)
+    from tvm.contrib import graph_runtime
+    m = graph_runtime.create(graph, lib, ctx)
+    m.set_input(**params)
+    m.set_input('input', data)
+    m.run()
+    tvm_out=m.get_output(0)
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_bn_trainingmod.py]: space between `=`

File: tests/python/frontend/tensorflow/test_bn_trainingmod.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    tvm.testing.assert_allclose(tvm_out.asnumpy(), tf_out.astype(tvm_out.dtype), rtol=1e-3)
+
+if __name__ == "__main__":
+    test_fusedbatchnorm()
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_bn_trainingmod.py]: test_fused_batch_norm

File: tests/python/frontend/tensorflow/test_bn_trainingmod.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+BatchNorm without given mean and variance given testcases
+====================
+This article is a test script to test fused_batch_norm operators in TensorFlow frontend when mean and variance are not given.
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_bn_trainingmod.py]: remove `article`

File: python/tvm/relay/frontend/tensorflow.py [hunk 7]
```diff
@@ -887,7 +887,14 @@ def _impl(inputs, attr, params):
         if 'U' in attr:
             need_cast = True
             inputs[0] = _op.cast(inputs[0], dtype=attr['U'].name)
-
+        # Check if mean and variance are empty
+        # If so, replace them with Mean and Variance Ops
+        # For run-time calculation
+        moving_mean_shape = [int(n) for n in inputs[3].type_annotation.shape]
```
Comment (FrozenGene) [inline python/tvm/relay/frontend/tensorflow.py]: It is the time to add `CHECK` of `len(inputs)`

Comment (lfengad) [reply python/tvm/relay/frontend/tensorflow.py]: Thank you for the detailed review comments! I have corrected the related codes in the newest commit. But for the `add CHECK of len(inputs)` issue, does it mean that to check the number of nodes in `inputs` and to see whether this number is 5? Also, it seems in python there is no `CHECK` method?
Thank you so much for the help!

Comment (FrozenGene) [reply python/tvm/relay/frontend/tensorflow.py]: The `CHECK` means `assert`. 

Yes, `CHECK` the number of nodes, so that we could visit `inputs[3]` / `inputs[4]` and so on safely.

File: tests/python/frontend/tensorflow/test_bn_dynamic.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        out = tf.identity(bn[0], name='sum')
+    data = np.random.rand(1, 12, 12, 32)
+    with tf.Session(graph=out.graph) as sess:
+        sess.run([tf.global_variables_initializer()])
+        tf_out = sess.run(out, feed_dict={input_tensor:data})
+        constant_graph = graph_util.convert_variables_to_constants(sess, sess.graph_def, ['sum'])
+
+    layout = None
+    target = 'llvm'
+    ctx = tvm.cpu(0)
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_bn_dynamic.py]: change to the code like this so that we could pass test when we don't enable `llvm`:
```python
        for device in ["llvm"]:
            ctx = tvm.context(device, 0)
            if not ctx.exist:
                print("Skip because %s is not enabled" % device)
                continue
```

File: python/tvm/relay/frontend/tensorflow.py (hunk 7 above)
Comment (lfengad) [reply python/tvm/relay/frontend/tensorflow.py]: I see. Thank you so much!

File: tests/python/frontend/tensorflow/test_bn_dynamic.py (hunk 8 above)
Comment (lfengad) [reply tests/python/frontend/tensorflow/test_bn_dynamic.py]: I have already modified the related code according to your suggestions, as in the newest commit. Thank you so much for your help!

File: tests/python/frontend/tensorflow/test_bn_dynamic.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        tf_out = sess.run(out, feed_dict={input_tensor:data})
+        constant_graph = graph_util.convert_variables_to_constants(sess, sess.graph_def, ['output'])
+
+    for device in ["llvm"]:
+        ctx = tvm.context(device, 0)
+        if not ctx.exist:
+            print("Skip because %s is not enabled" % device)
+            continue
+        layout = None
+        mod, params = relay.frontend.from_tensorflow(constant_graph, layout=layout, outputs=['output'])
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_bn_dynamic.py]: I think `layout = None` is unnecessary. Because for CPU, we use the default `NHWC` layout is enough. The `layout` is used when we have `convolution` ops, which will be `NCHW`  better if target is `GPU`.

File: tests/python/frontend/tensorflow/test_bn_dynamic.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        with relay.build_config(opt_level=3):
+            graph, lib, params = relay.build(mod,
+                                             target=device,
+                                             target_host=None,
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_bn_dynamic.py]: `target_host = None` is unnecessary. Because our default value of it is `None`.

Comment (lfengad) [reply tests/python/frontend/tensorflow/test_bn_dynamic.py]: I have already modified the code as your suggestions. 
Thank you so much for the review!

File: tests/python/frontend/tensorflow/test_bn_dynamic.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        from tvm.contrib import graph_runtime
+        m = graph_runtime.create(graph, lib, ctx)
+        m.set_input(**params)
+        m.set_input('input', data)
+        m.run()
+        tvm_out = m.get_output(0)
+        tvm.testing.assert_allclose(tvm_out.asnumpy(), tf_out.astype(tvm_out.dtype), rtol=1e-3)
+
+if __name__ == "__main__":
+    test_fused_batch_norm()
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_bn_dynamic.py]: Sorry, I think I have another one comment. How about us add some more testing data?

for example:

```python
def verify_fused_batch_norm(shape):
      ....

def test_fused_batch_norm():
    verify_fused_batch_norm(shape=(1, 12, 12, 32))
    verify_fused_batch_norm(shape=(1, 24, 24, 64))
    ...

if __name__ == "__main__":
    test_fused_batch_norm()

Comment (lfengad) [reply tests/python/frontend/tensorflow/test_bn_dynamic.py]: Sure, I think such a test could be more convincing.

Comment (lfengad) [reply tests/python/frontend/tensorflow/test_bn_dynamic.py]: > Sorry, I think I have another one comment. How about adding some more testing data?
> 
> for example:
> 
> ```python
> def verify_fused_batch_norm(shape):
>     ...
> 
> def test_fused_batch_norm():
>     verify_fused_batch_norm(shape=(1, 12, 12, 32))
>     verify_fused_batch_norm(shape=(1, 24, 24, 64))
>     ...
> 
> if __name__ == "__main__":
>     test_fused_batch_norm()
> ```

I have modified the code as your suggestions to add more testing cases as in the newest commit.  
Thank you for the comments!

**Comment by FrozenGene** [comment]
Let us wait CI green.

As GitHub has issue: https://discuss.tvm.ai/t/github-issue-the-commit-author-is-wrong-since-today/5880/15 I will merge it after it is solved.

**Comment by lfengad** [comment]
> Let us wait CI green.
> 
> As GitHub has issue: https://discuss.tvm.ai/t/github-issue-the-commit-author-is-wrong-since-today/5880/15 I will merge it after it is solved.

Okay, thank you so much for the efforts!

File: tests/python/frontend/tensorflow/test_bn_dynamic.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+            graph, lib, params = relay.build(mod,
+                                             target=device,
+                                             params=params)
+        from tvm.contrib import graph_runtime
+        m = graph_runtime.create(graph, lib, ctx)
+        m.set_input(**params)
+        m.set_input('input', data)
+        m.run()
+        tvm_out = m.get_output(0)
+        tvm.testing.assert_allclose(tvm_out.asnumpy(), tf_out.astype(tvm_out.dtype), rtol=1e-1)
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_bn_dynamic.py]: Why change to 1e-1? atol=1e-3, rtol=1e-3 can not satisify?

Comment (lfengad) [reply tests/python/frontend/tensorflow/test_bn_dynamic.py]: Yeah, 1e-3 cannot be satisfied from the CI. The max related difference from the previous CI run is about 0.007.

Comment (FrozenGene) [reply tests/python/frontend/tensorflow/test_bn_dynamic.py]: I doubt it is not true. Could you double check it? You could refer our exist testing case. atol=1e3 rtol=1e-3 should work. Your code omit `atol`

Comment (lfengad) [reply tests/python/frontend/tensorflow/test_bn_dynamic.py]: Okay, I can try it. Thank you for the suggestions!

Comment (lfengad) [reply tests/python/frontend/tensorflow/test_bn_dynamic.py]: Yeah, it works now! I didn't realize the atol attribute. The default setting of atol is too small. Sorry about that!

**Comment by FrozenGene** [comment]
Thanks @lfengad This is merged now.

**Comment by lfengad** [comment]
> Thanks @lfengad This is merged now.

Thank you so much for your help! 😄
